### PR TITLE
Add OpenMM recipe (CPU only) [WIP]

### PR DESCRIPTION
**Pull request by jaimergp**
Draft recipe following #9102. This will provide `openmm` with CPU-only capabilities (no GPU).

## Checklist

- [X] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [X] Source is from official source
- [X] Package does not vend other packages
- [X] Build number is 0
- [X] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by Lnaden** [comment]
I would like to offer one bit of caution with this. Because of how this package is deployed already (though Omnia), if you ever upload a package to Conda-Forge called "openmm", it will take priority over the `omnia` channel since we require all our users to specify `conda-forge` as a higher priority channel.

The whole point of that of course was to move people over to conda-forge when GPU support was available, but just a word of caution in case you try to upload something called "openmm" before its ready to supersede the `omnia` channel one.

That said, I'm really excited to see this come to Conda-Forge

**Comment by Lnaden** [comment]
And yes, I see this is a package called "openmm-cpu", so its okay here, just a heads up.

**Comment by jaimergp** [comment]
Oh, yes, thanks for the heads up. That could have totally happened. I'll keep an eye for this in #9102!

File: recipes/openmm-cpu/conda_build_config.yaml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+python:
```
Comment (marcelotrevisani) [inline recipes/openmm-cpu/conda_build_config.yaml]: you do not need this file, you just have to select those pythons which you want to in your recipe

File: recipes/openmm-cpu/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  name: openmm-cpu
+  version: {{ version }}
+
+source:
+  url: https://github.com/openmm/{{ name }}/archive/{{ version }}.tar.gz
+  sha256: {{ checksum }}
+
+
+build:
+  number: 0
```
Comment (marcelotrevisani) [inline recipes/openmm-cpu/meta.yaml]:
```suggestion
  number: 0
  skip: true  # [py2k]
```

Comment (marcelotrevisani) [reply recipes/openmm-cpu/meta.yaml]: With this modification you can skip Python 2

File: recipes/openmm-cpu/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - jom  # [win]
+
+  host:
+    - python  {{ python }}
+    - fftw
+    - swig
+    - doxygen ==1.8.14
```
Comment (marcelotrevisani) [inline recipes/openmm-cpu/meta.yaml]:
```suggestion
    - doxygen 1.8.14
```
I think you can drop the docs build, or do you really need want/need them?

File: recipes/openmm-cpu/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    # - sphinxcontrib-lunrsearch >=0.3
+    # - sphinxcontrib-autodoc_doxygen
+    # # Required for sphinx
+    # - latexmk  # [not win]
+    - lxml
+    - numpy
+    - cython
+
+  run:
+    - python  {{ python }}
```
Comment (marcelotrevisani) [inline recipes/openmm-cpu/meta.yaml]:
```suggestion
    - python
```

File: recipes/openmm-cpu/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - jom  # [win]
+
+  host:
+    - python  {{ python }}
```
Comment (marcelotrevisani) [inline recipes/openmm-cpu/meta.yaml]:
```suggestion
    - python
```

File: recipes/openmm-cpu/conda_build_config.yaml (hunk 1 above)
Comment (jaimergp) [reply recipes/openmm-cpu/conda_build_config.yaml]: Thanks for the suggestion! Applied!

File: recipes/openmm-cpu/meta.yaml (hunk 3 above)
Comment (jaimergp) [reply recipes/openmm-cpu/meta.yaml]: I'll check with the authors whether they want them bundled.

**Comment by jaimergp** [comment]
@conda-forge-admin please rerender

**Comment by jaimergp** [comment]
@conda-forge/core, how can I...?

- Disable Appveyor builds? (Windows is already being built on Azure)
- Enable `ppc64le` builds in Azure?

Basically, I want to _rerender_ this fork with the configuration specified at `conda-forge.yml`, but I don't think this is supported, right?

**Comment by jakirkham** [comment]
@Lnaden, have you seen the work we've been doing on the GPU front? 😉

ref: https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/237

**Comment by Lnaden** [comment]
I have and I am very excited to see that come to fruition and get fully documented for users to build against!

**Comment by bgruening** [comment]
@jaimergp I think we can ignore AppVeyor here as it was built on Azure just fine. If you take this out of Draft we can review and merge it and you rerender it in your feedstock.

File: recipes/openmm-cpu/build.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+fi
+
+
+# Set location for FFTW3 on both linux and mac
+CMAKE_FLAGS+=" -DFFTW_INCLUDES=${PREFIX}/include/"
+CMAKE_FLAGS+=" -DFFTW_LIBRARY=${PREFIX}/lib/libfftw3f${SHLIB_EXT}"
+CMAKE_FLAGS+=" -DFFTW_THREADS_LIBRARY=${PREFIX}/lib/libfftw3f_threads${SHLIB_EXT}"
+
+
+# DEBUG
```
Comment (bgruening) [inline recipes/openmm-cpu/build.sh]: can be removed?

File: recipes/openmm-cpu/build.sh [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+echo "CMAKE_FLAGS:"
+echo $CMAKE_FLAGS
+
+
+# Build in subdirectory and install.
+mkdir build
+cd build
+cmake ${CMAKE_FLAGS} ${SRC_DIR}
+make -j$CPU_COUNT
+make -j$CPU_COUNT install PythonInstall
```
Comment (bgruening) [inline recipes/openmm-cpu/build.sh]: is there a make test target?

File: recipes/openmm-cpu/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  version: {{ version }}
+
+source:
+  url: https://github.com/openmm/{{ name }}/archive/{{ commit }}.tar.gz
+  sha256: 720c9aa0215e2dd851c40a6319814869cb55aa82ef908476c76bec3d494dec1d
+
+
+build:
+  number: {{ build }}
+  string: "rc_py{{ py }}_h{{ PKG_HASH }}_{{ build }}"
```
Comment (bgruening) [inline recipes/openmm-cpu/meta.yaml]: Why is this string needed?

**Comment by jaimergp** [comment]
@bgruening Once the feedstock is ready, would that mean `openmm-cpu` would be automatically uploaded to the `conda-forge` channel? If that's the case, we might need to wait on some feedback by the OpenMM developers. If there is a way we can control when to ship the package, then I'd love to have the feedstock to test `ppc64le`.

Pinging @jchodera and @peastman here to participate in the review.

File: recipes/openmm-cpu/meta.yaml (hunk 8 above)
Comment (jaimergp) [reply recipes/openmm-cpu/meta.yaml]: I am adapting the original [`omnia` recipe](https://github.com/omnia-md/conda-recipes-cuda/blob/master/openmm/meta.yaml#L22), and also read about it in the [docs](https://conda-forge.org/docs/maintainer/knowledge_base.html#preferring-a-provider-usually-nompi). Since it was being used to specify the CUDA version in the [sibling recipe](https://github.com/conda-forge/staged-recipes/pull/9102), it ended up looking like this. Of course, please suggest the best practices here. @jchodera might want to add something too.

File: recipes/openmm-cpu/build.sh (hunk 7 above)
Comment (jaimergp) [reply recipes/openmm-cpu/build.sh]: I _think_ there is one, but it takes a while to finish. @peastman might want to add something here.

File: recipes/openmm-cpu/build.sh (hunk 6 above)
Comment (jaimergp) [reply recipes/openmm-cpu/build.sh]: Yes, thank you.

File: recipes/openmm-cpu/build.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+# Set location for FFTW3 on both linux and mac
+CMAKE_FLAGS+=" -DFFTW_INCLUDES=${PREFIX}/include/"
+CMAKE_FLAGS+=" -DFFTW_LIBRARY=${PREFIX}/lib/libfftw3f${SHLIB_EXT}"
+CMAKE_FLAGS+=" -DFFTW_THREADS_LIBRARY=${PREFIX}/lib/libfftw3f_threads${SHLIB_EXT}"
+
+
+# DEBUG
+env | sort
+echo "CMAKE_FLAGS:"
+echo $CMAKE_FLAGS
```
Comment (jaimergp) [inline recipes/openmm-cpu/build.sh]: @bgruening You mean from `#DEBUG` to this line, right?

**Comment by simonbray** [comment]
Hey @jaimergp, how are you progressing? This looks almost ready to merge to me.

File: recipes/openmm-cpu/build.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+if [[ "$OSTYPE" == "linux-gnu" ]]; then
+    # CFLAGS
+    # JRG: Had to add -ldl to prevent linking errors (dlopen, etc)
+    MINIMAL_CFLAGS+=" -g -O3 -ldl"
+    CFLAGS+=" $MINIMAL_CFLAGS"
+    CXXFLAGS+=" $MINIMAL_CFLAGS"
+    LDFLAGS+=" $LDPATHFLAGS"
+
+    # From https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#an-aside-on-cmake-and-sysroots
+    CMAKE_FLAGS+=" -DCMAKE_TOOLCHAIN_FILE=${RECIPE_DIR}/cross-linux.cmake"
```
Comment (isuruf) [inline recipes/openmm-cpu/build.sh]: This is unnecessary.

File: recipes/openmm-cpu/build.sh [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+# Ensure we build a release
+CMAKE_FLAGS+=" -DCMAKE_BUILD_TYPE=Release"
+
+if [[ "$OSTYPE" == "linux-gnu" ]]; then
+    # CFLAGS
+    # JRG: Had to add -ldl to prevent linking errors (dlopen, etc)
+    MINIMAL_CFLAGS+=" -g -O3 -ldl"
+    CFLAGS+=" $MINIMAL_CFLAGS"
+    CXXFLAGS+=" $MINIMAL_CFLAGS"
```
Comment (isuruf) [inline recipes/openmm-cpu/build.sh]: Dont override. Append

File: recipes/openmm-cpu/build.sh [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    # From https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#an-aside-on-cmake-and-sysroots
+    CMAKE_FLAGS+=" -DCMAKE_TOOLCHAIN_FILE=${RECIPE_DIR}/cross-linux.cmake"
+    # Use GCC
+    CMAKE_FLAGS+=" -DCMAKE_C_COMPILER=$CC -DCMAKE_CXX_COMPILER=$CXX"
+
+
+elif [[ "$OSTYPE" == "darwin"* ]]; then
+    # conda-build MACOSX_DEPLOYMENT_TARGET must be exported as an environment variable to override 10.7 default
+    # cc: https://github.com/conda/conda-build/pull/1561
+    export MACOSX_DEPLOYMENT_TARGET="10.13"
```
Comment (isuruf) [inline recipes/openmm-cpu/build.sh]: Don't set this. conda-forge sets this value

File: recipes/openmm-cpu/build.sh (hunk 11 above)
Comment (jaimergp) [reply recipes/openmm-cpu/build.sh]: Isn't `+=` the append to string operator?

File: recipes/openmm-cpu/build.sh [hunk 13]
```diff
@@ -0,0 +1,53 @@
+#!/bin/bash
+
+CMAKE_FLAGS="-DCMAKE_INSTALL_PREFIX=$PREFIX -DBUILD_TESTING=OFF"
+
+# Ensure we build a release
+CMAKE_FLAGS+=" -DCMAKE_BUILD_TYPE=Release"
+
+if [[ "$OSTYPE" == "linux-gnu" ]]; then
```
[the remark this answers is not in the document]
Comment (jaimergp) [reply recipes/openmm-cpu/build.sh]: Nice, thanks! Which is the equivalent for MacOS?

[the remark this answers is not in the document]
Comment (isuruf) [reply recipes/openmm-cpu/build.sh]: "$target_platform" == osx*

File: recipes/openmm-cpu/build.sh (hunk 11 above)
Comment (isuruf) [reply recipes/openmm-cpu/build.sh]: Yes, you are right

**Comment by jaimergp** [comment]
Hi @simonbray!

I am mostly done! I will test and apply @isuruf's suggestions now, but we are still waiting for @peastman and @jchodera's feedback on some questions opened by @bgruening.

**Comment by jaimergp** [comment]
Also, can someone clarify this?

> Once the feedstock is ready, would that mean `openmm-cpu` would be automatically uploaded to the `conda-forge` channel? If that's the case, we might need to wait on some feedback by the OpenMM developers. If there is a way we can control when to ship the package, then I'd love to have the feedstock to test `ppc64le`.

**Comment by jaimergp** [comment]
Oh, and last one, I promise. Windows builds show [some warnings (line 10866)](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=63435&view=logs&jobId=240f1fee-52bc-5498-a14a-8361bde76ba0&taskId=23010c01-6a9d-5f94-9037-792618644b77&lineStart=10866&lineEnd=10867&colStart=1&colEnd=1) about (over)linking because `vs2015_runtime` is not specified as a dependency. Should I add it somewhere? I thought this would be handled by the `{{ compiler() }}` magic.

**Comment by simonbray** [comment]
Yes, once the feedstock is created the package is automatically built and uploaded. I suppose it would be possible to suppress this process, but I haven't seen that done before. It might be better to keep development in this PR until you feel the package is ready to publish.

File: recipes/openmm-cpu/build.sh (hunk 7 above)
Comment (peastman) [reply recipes/openmm-cpu/build.sh]: Yes, that's correct.  If you're only building the Reference and CPU platforms, `make test` should only take a few minutes to run.  Testing CUDA and OpenCL takes much longer.  Be aware there are some stochastic tests that occasionally fail even when everything is working correctly, so you need to be able to tolerate that.  On travis we have it automatically rerun any tests that fail and only report a problem if a test fails twice.

Comment (jchodera) [reply recipes/openmm-cpu/build.sh]: We probably don't want to run the tests during the recipe build though, since running all the unit tests takes a very long time.

Instead, we normally run a quick Python test to make sure everything was built properly and the platforms perform as expected:
```bash
python openmm/wrappers/python/simtk/testInstallation.py
```
Perhaps we could use that here?

File: recipes/openmm-cpu/meta.yaml (hunk 8 above)
Comment (jchodera) [reply recipes/openmm-cpu/meta.yaml]: I don't think we need a build string here, and possibly not even in the CUDA recipe either. 

Previously, on the `omnia` build infrastructure, we were building different variants for each CUDA Toolkit version. For conda-forge, for GPU-enabled OpenMM, we can deliver a specific `cudatoolkit`as a dependency and make sure OpenMM gets compiled against that version, so we may not need build strings there either.

File: recipes/openmm-cpu/build.sh (hunk 7 above)
Comment (jchodera) [reply recipes/openmm-cpu/build.sh]: Actually, we already do this:
https://github.com/conda-forge/staged-recipes/pull/9163/files#diff-d602ab125553c5093251c27f0a2fe342R52

Comment (jaimergp) [reply recipes/openmm-cpu/build.sh]: I'll leave it as is, then!

File: recipes/openmm-cpu/meta.yaml (hunk 8 above)
Comment (jaimergp) [reply recipes/openmm-cpu/meta.yaml]: Removing the `string` line for now.

**Comment by bgruening** [comment]
Is this ready for review?

**Comment by jaimergp** [comment]
Commit 654231b fixed all Windows warnings, both in Azure AND AppVeyor!

The main difference is in [these lines](https://github.com/conda-forge/staged-recipes/pull/9163/files#diff-935cef312aaf26d7ddfa44b0d7f607ecR19-R21). The original recipe left the compiled libraries at `PREFIX/Lib/site-packages/simtk/openmm/` and `PREFIX/Library/lib/`, but apparently they should be at `PREFIX/Library/bin`, so I am just copying them. It's not the cleanest approach, but I'd say it's safe and it's working!

I am adding @jchodera and @peastman as maintainers. Please confirm whether you are ok with this! Once you agree, I'd say we can merge this.

**Comment by jaimergp** [comment]
Well... I might have spoken too soon.

While Azure [sees CPU and Reference platforms](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=71901&view=logs&jobId=240f1fee-52bc-5498-a14a-8361bde76ba0&taskId=23010c01-6a9d-5f94-9037-792618644b77&lineStart=9268&lineEnd=9269&colStart=1&colEnd=1), Appveyor only sees [Reference](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/27393976#L9700).

<details>

<summary>Azure</summary>

```
(%PREFIX%) %SRC_DIR%>python -m simtk.testInstallation 

OpenMM Version: 7.4
Git Revision: Unknown

There are 2 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.30498e-06

All differences are within tolerance.
``` 

</details>

<details>

<summary>Appveyor</summary>

```
(%PREFIX%) %SRC_DIR%>python -m simtk.testInstallation 
OpenMM Version: 7.4
Git Revision: Unknown
There are 1 Platforms available:
1 Reference - Successfully computed forces
All differences are within tolerance.
```
</details>


Any ideas, @peastman or @jchodera?

**Comment by jchodera** [comment]
Can you try
```python
from simtk import openmm
openmm.Platform.getPluginLoadFailures()
```

**Comment by peastman** [comment]
What kind of CPU is it using?  The only reason I know of that the CPU platform wouldn't load is if the processor doesn't support SSE 4.1, which is only true of *very* old ones.

**Comment by jaimergp** [comment]
> Can you try
> 
> ```python
> from simtk import openmm
> openmm.Platform.getPluginLoadFailures()
> ```

```
(base) C:\Users\Jaime>conda activate openmm

(openmm) C:\Users\Jaime>python -m simtk.testInstallation

OpenMM Version: 7.4
Git Revision: Unknown

There are 1 Platforms available:

1 Reference - Successfully computed forces

All differences are within tolerance.

(openmm) C:\Users\Jaime>python
Python 3.6.7 (default, Jul  2 2019, 02:21:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from simtk import openmm
>>> openmm.Platform.getPluginLoadFailures()
()
```

Interestingly, I am able to reproduce this if I build the recipe in my Windows machine, so I don't think it's due to the SSE support.

**Comment by peastman** [comment]
Is the CPU platform getting built?  Look in the plugins directory.  Are the OpenMMCPU and OpenMMPME libraries there?

**Comment by jaimergp** [comment]
I'd say it's being built because it appears in the [linking review](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/27396215?fullLog=true#L9572), but it might happen that it cannot be found because I didn't copy the DLLs under `plugins` to `Library/bin`. Let me check if that will solve it.

**Comment by jaimergp** [comment]
Well, no luck. The `*CPU` libs are getting built [for sure](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/27459510#L9385), but we still see a single platform on AppVeyor...

**Comment by peastman** [comment]
> it might happen that it cannot be found because I didn't copy the DLLs under plugins to Library/bin

That probably isn't the right location.  Try this.

```python
from simtk.openmm import *
print(version.openmm_library_path)
```

The directory it prints should contain a subdirectory called `plugins`.  That's where the plugin libraries should be.

**Comment by jaimergp** [comment]
Copying the DLLs to `Library/bin` solved the warnings _and_ fixed Appveyor builds (which didn't succeed at all before that). It's also what `conda-build` seems to [recommend](https://docs.conda.io/projects/conda-build/en/latest/resources/use-shared-libraries.html?highlight=shared%20libraries#shared-libraries-in-windows). Anyway, I am just copying them, so they are also in the original location (`%PREFIX%/Library/lib/`, which contains `plugins`). Do they need to reside under their own subdirectory, i.e. `%PREFIX%/Library/bin/plugins`?

Notice that I am using `%LIBRARY_PREFIX%` in Windows (instead of  `%PREFIX%`, the [current approach](https://github.com/omnia-md/conda-recipes-cuda/blob/master/openmm/bld.bat#L7)) because I am assuming that OpenMM does indeed follow a Unix-like conventions for the package structure, as recommended [here](https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html?highlight=environment%20variables#environment-variables-set-during-the-build-process) (2nd table).

Anyway, the location must be at least partially right because Azure does see both platforms... :man_shrugging:

**Comment by peastman** [comment]
> Do they need to reside under their own subdirectory, i.e. %PREFIX%/Library/bin/plugins?

It loads whatever plugin libraries it finds in `os.path.join(version.openmm_library_path, 'plugins')`.

**Comment by jaimergp** [comment]
Last commit didn't fix it. I will try to debug it further tomorrow. It still puzzles me _why_ it does work on Azure...

**Comment by peastman** [comment]
Clearly there's something different.  Anyway, since you're able to reproduce it locally that should be the easy way to debug.  `version.openmm_library_path` tells you where it's looking for the `plugins` directory.  Look there to see whether the plugins are present.  If they are, `Platform.getPluginLoadFailures()` will tell you if any of them couldn't be loaded.

**Comment by jaimergp** [comment]
We would expect [this line](https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/27476353#L9765) to raise an `AssertionError` if there were any load failures, right?

```
python -c "import simtk.openmm as mm; err = mm.Platform.getPluginLoadFailures(); assert not err, err" 
```

I'll check locally in my Windows box tomorrow.

**Comment by peastman** [comment]
I think that ought to do it.

**Comment by jaimergp** [comment]
In my Windows installation:

```
(openmm) C:\Users\Jaime>python
Python 3.6.7 (default, Jul  2 2019, 02:21:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from simtk.openmm import version
'C:/bld/openmm-cpu_1568381014179/_h_env/Library/lib'
```

This is the path used during the build process, not the actual installation path (`C:\Users\Jaime\Miniconda3\envs\openmm\Library\lib`).

If I manually set `OPENMM_PLUGIN_DIR`, voilà:

```
(openmm) C:\Users\Jaime\>set OPENMM_PLUGIN_DIR=C:\Users\Jaime\Miniconda3\envs\openmm\Library\lib\plugins

(openmm) C:\Users\Jaime\>python -m simtk.testInstallation

OpenMM Version: 7.4
Git Revision: Unknown

There are 2 Platforms available:

1 Reference - Successfully computed forces
2 CPU - Successfully computed forces

Median difference in forces between platforms:

Reference vs. CPU: 6.28968e-06

All differences are within tolerance.
```

So the question is... how can I fix that location in the recipe?

**Comment by jaimergp** [comment]
`simtk.openmm.version` has these contents:

```
# THIS FILE IS GENERATED FROM OPENMM SETUP.PY
short_version = '7.4.0'
version = '7.4.0'
full_version = '7.4.0.dev-Unknown'
git_revision = 'Unknown'
release = False
openmm_library_path = r'C:/bld/openmm-cpu_1568795804938/_h_env/Library/lib'

if not release:
    version = full_version
```

`conda-build` should recognize this hardcoded path and replace it, but for some reason it's not doing that. I will push a commit now to see if this is different in Azure.

**Comment by jaimergp** [comment]
Yes, Azure does recognize the hardcoded path:

```
WARNING: Detecting which files contain PREFIX is slow, installing ripgrep makes it faster. 'conda install ripgrep'
Detected hard-coded path in text file Lib/site-packages/simtk/openmm/version.py
```

These lines do not appear in the Appveyor logs.

**Comment by jaimergp** [comment]
Ok, finally got it! It has to do with `ripgrep`. If the `rg` executable is found on the system, `conda-build` [will use it](https://github.com/conda/conda-build/blob/master/conda_build/build.py#L186-L216) to accelerate the hardcoded  `PREFIX` searches. 

However, for some reason `simtk.openmm.version` is not found by the `rg` approach, so it ends up being ignored in the search. If I run `conda remove --force ripgrep` on the base environment on my Windows machine, then `rg` is not found and `conda-build` must go the slow way, which does find `version.py`! In that case, the hardcoded `PREFIX` gets replaced, the library path is correct, the plugins are loaded and the two platforms are found.

In Appveyor, `ripgrep` is used; in Azure, it's not (although is definitely installed!).

This leads to two issues:

1. `conda-forge` issue: Why is not Azure using `rg` if `ripgrep` is [present](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=73158&view=logs&jobId=240f1fee-52bc-5498-a14a-8361bde76ba0&taskId=cc9510cc-c1c4-564b-b6ee-50804dbb3143&lineStart=85&lineEnd=86&colStart=1&colEnd=1) in the base environment?
2. `conda-build` issue: Why does not `ripgrep` find our offending file?

To sum up, I would say that we can __safely ignore the errors on the AppVeyor build__ because it's a bug on the conda-build side that it's not currently affecting Azure.

_I will submit bug reports for both issues now._

**Comment by jchodera** [comment]
Woohoo!

**Comment by jaimergp** [comment]
Just confirmed on Appveyor! The `ripgrep` detection is indeed the one to blame. I will clean the recipes a bit from all those debugging lines and will mark this as _ready for review_. We only want to use Azure anyway.

I guess by now both @peastman and @jchodera are OK with being maintainers, right? I think we need written confirmation (i.e. a comment saying so) before proceeding.

**Comment by jchodera** [comment]
I'm OK with being a maintainer!

**Comment by peastman** [comment]
Fine with me.

**Comment by jaimergp** [comment]
@conda-forge/staged-recipes, this is finally ready for review!

Please, ignore Appveyor errors as stated [here](https://github.com/conda-forge/staged-recipes/pull/9163#issuecomment-532652100).

**Comment by jaimergp** [comment]
One question to be addressed during the review:

This package `openmm-cpu` is a CPU-only build of `openmm` (which is currently being build in the `omnia` channel, with CPU/CUDA/OpenCL support). I guess that we don't want users to install both `openmm` and `openmm-cpu` at the same time. Is [this `run_constrained` trick](https://conda-forge.org/docs/maintainer/adding_pkgs.html#defining-conflicts) the way to go?

**Comment by jchodera** [comment]
Note that we're hoping to also build a version of OpenMM with GPU support in `conda-forge` as soon as we can figure out how to compile it against the CUDA toolkit using the conda-forge build infrastructure!

**Comment by jakirkham** [comment]
Just recently got one feedstock to build GPU packages. It also does CPU only packages in the same feedstock. Maybe this is a good template for you? Here's the PR ( https://github.com/conda-forge/ucx-split-feedstock/pull/14 ). Please let me know if you have any questions 🙂

**Comment by jaimergp** [comment]
@jakirkham Thanks for the heads up! Sorry to ask but, how does the metapackage work exactly? I can't see how a user would configure their preference regarding GPU vs CPU.

**Comment by jakirkham** [comment]
They would run `conda install ucx-proc=*=gpu` or similar for `cpu`. This could be done in conjunction with other package installs.

**Comment by jakirkham** [comment]
Maybe there are other ways to accomplish this task. I'm not really sure. There is precedent for using metapackages for analogous problems like MPI and BLAS. So it seemed reasonable to do the same thing here.

We haven't dealt with a default variant in the `ucx` case. We could use features to down prioritize one of them. It's hard for me to say what a good prioritization would be in our case at least. So we've left it up to the user to make a choice.

It might be worth breaking out this discussion into a new issue on the webpage repo. Perhaps other people have thoughts on these questions. What do you think?

**Comment by scopatz** [comment]
> Perhaps other people have thoughts on these questions. What do you think?

I think metapackages make sense here

**Comment by jaimergp** [comment]
We won't be building a CPU-only _package_ in the end (maybe a variant, if any), so I'll close this PR. Thanks for all the comments and feedback! Focus is now on #9102.
